### PR TITLE
TCPStore watchKey and callback cpp tests

### PR DESCRIPTION
**Pull request by H-Huang**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#54734 TCPStore watchKey and callback cpp tests**
* #54264 TCPStore add watchKey method and new listener thread

**Changes**

- Clang-format the existing tests. 

- Added new test which tests `watchKey(key, callback)` by spinning up multiple threads which execute `set(key, ...)`, `compareSet(key, ...)`, and `add(key, ...)`. 
These all modify the `key` and should trigger a callback execution. Then we ensure that the number of those method calls is equal to number of times the callback was actually executed.

### CODE REVIEW

**Comment by pritamdamania87** [comment]
Code and tests for a feature should be in the same PR :)

File: torch/lib/c10d/test/TCPStoreTest.cpp [hunk 1]
```diff
@@ -136,10 +142,89 @@ void testHelper(const std::string& prefix = "") {
...
+  std::vector<c10::intrusive_ptr<c10d::PrefixStore>> clientStores;
+  for (auto i = 0; i < numThreads; i++) {
+    clientTCPStores.push_back(c10::make_intrusive<c10d::TCPStore>(
+        "127.0.0.1", serverTCPStore->getPort(), numWorkers, false));
+    clientStores.push_back(
+        c10::make_intrusive<c10d::PrefixStore>(prefix, clientTCPStores[i]));
+  }
+
+  std::vector<std::thread> threads;
+  int keyChangeOperationCount = 0;
```
Comment (pritamdamania87) [inline torch/lib/c10d/test/TCPStoreTest.cpp:186]: This needs to be atomic?

File: torch/lib/c10d/test/TCPStoreTest.cpp [hunk 2]
```diff
@@ -136,10 +142,89 @@ void testHelper(const std::string& prefix = "") {
   }
 }
 
+void testWatchKeyCallback(const std::string& prefix = "") {
+  // Callback function increments counter of the total number of callbacks that
+  // were run
+  int numCallbacksExecuted = 0;
+  std::function<void(std::string, std::string)> callback =
+      [&numCallbacksExecuted](
+          const std::string& old_value, const std::string& new_value) {
```
Comment (pritamdamania87) [inline torch/lib/c10d/test/TCPStoreTest.cpp:151]: We should also validate we are receiving the right `old_value` and `new_value` in the callbacks as well.

**Comment by H-Huang** [comment]
Combined these changes with previous PR

**Pull request closed without merging**